sympy/core/evalf.py

```diff
@@ -81,9 +101,16 @@ class PrecisionExhausted(ArithmeticError):
 if the corresponding complex part is None.
 
 """
+TMP_RES = Any  # temporary result, should be some variant of
+# tUnion[tTuple[Optional[MPF_TUP], Optional[MPF_TUP], Optional[int], Optional[int]], 'ComplexInfinity']
+# but mypy reports error because it doesn't know as we know
+# 1. re and re_acc are either both None or both MPF_TUP
+# 2. sometimes the result can't be zoo
+
+OPT_DICT = tDict[str, Any]  # type of the "options" parameter in internal evalf functions
 
 
-def fastlog(x):
+def fastlog(x: Optional[MPF_TUP]) -> tUnion[int, Any]:
```

Comment:
Is the return here not `tUnion[float, int]`?

Reply:
Declaring float return type will cause many operations to generate float value

Reply:
I don't understand what you mean. The type annotations do not affect the behaviour of the code so I guess you mean that in some cases the type is inferred to be float but I'm not sure how that's a problem given that the return type can be a float.

My reading of the function is that it either returns `MINUS_INF` which is a float or it returns `x[2]+x[3]` where `x` has type `MPF_TUP` so that's a sum of ints which gives an int.

Reply:
https://github.com/sympy/sympy/blob/dfd966437a2ac0551c2c55174187f92dd6cba278/sympy/core/evalf.py#L275-L291
When line 283 is executed, we already know `re` is neither `fzero` nor `None`, so `size_re` should be int, but mypy doesn't know that. Here is an example that mypy reports error:
```py
def f(x: int) -> Union[int, None]:
    if x == 0:
        return None
    return x

a: int = f(3)
print(a, type(a))
```
Maybe `Literal` can be used to solve this problem, but that requires python 3.8
Or I can simply remove these kinds of return type hint if you prefer so.

Reply:
I don't actually know what I prefer. I'm basically learning how type hints work in practice as I go here (hence all my questions).